Make sure all BIGNUM operations work within the FIPS provider

mattcaswell opened the pull request:
The FIPS provider does not have a default OPENSSL_CTX so, where
necessary, we need to ensure we can always access an explicit
OPENSSL_CTX. We remove functions from the FIPS provider that use
the default OPENSSL_CTX, and fixup some places which were using
those removed functions.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

levitte commented on crypto/bn/bn_gf2m.c:
```diff
@@ -732,8 +732,8 @@ int BN_GF2m_mod_inv(BIGNUM *r, const BIGNUM *a, const BIGNUM *p, BN_CTX *ctx)
-        if (!BN_priv_rand(b, BN_num_bits(p) - 1,
-                          BN_RAND_TOP_ANY, BN_RAND_BOTTOM_ANY))
+        if (!BN_priv_rand_ex(b, BN_num_bits(p) - 1,
+                          BN_RAND_TOP_ANY, BN_RAND_BOTTOM_ANY, ctx))
```
Nit: indentation

levitte commented on crypto/bn/bn_prime.c:
```diff
@@ -84,41 +84,38 @@ int BN_GENCB_call(BN_GENCB *cb, int a, int b)
-                         const BIGNUM *add, const BIGNUM *rem, BN_GENCB *cb)
+int BN_generate_prime_ex2(BIGNUM *ret, int bits, int safe,
+                         const BIGNUM *add, const BIGNUM *rem, BN_GENCB *cb,
+                         BN_CTX *ctx)
```
Nit: indentation

slontis commented on crypto/bn/bn_prime.c:
```diff
@@ -171,23 +168,41 @@ int BN_generate_prime_ex(BIGNUM *ret, int bits, int safe,
 {
     int i, status, ret = -1;
-    BN_CTX *ctx = NULL;
+#ifndef FIPS_MODE
```
OK this is kind of scaring me.
Why do you allow this in non FIPS?
I have a concern that this is going to bite us at some point - especially since we are not running the 
tests using the FIPS provider yet..

t8m replied:
Yes, at least there is a missing if (ctx == NULL) goto err; in FIPS_MODE

levitte replied:
I'm not sure I understand why it's disabled for FIPS. Is there some weird consensus that BN_CTX shouldn't be used in the FIPS module?

t8m replied:
The BN_CTX needs to be passed because the FIPS module needs the non-default OPENSSL_CTX (which is inside the passed-in BN_CTX). The goal is quite clear.

What is not clear to me (but that might be because I did not study the provider implementation plan well enough) how this will all work from the application side - if the application calls BN_generate_prime_ex2 by itself what will happen? As I understand it now the default implementation be called even with FIPS provider being loaded?

levitte replied:
If the application calls BN code itself, then it's getting to the libcrypto build, not the FIPS build.

t8m replied:
Then it would still be beneficial to somehow guard against misuse of these APIs inside the FIPS provider itself. Coverity scans will hopefully reveal this but could we make it more explicit such as with __attribute__((nonnull)) annotations? However they would have to be FIPS_MODE only.

mattcaswell replied:
Not sure why this is considered scary. I could put in the `if (ctx == NULL) goto err;` line for FIPS_MODE if that lessens your worries?

> Why do you allow this in non FIPS?

Outside of FIPS_MODE the presence of a `BN_CTX` is simply an optimisation. It allows the reuse of temporary BIGNUM objects. Inside FIPS_MODE it must be present because we get the current OPENSSL_CTX from it. Since FIPS_MODE requires an explicit OPENSSL_CTX (the default OPENSSL_CTX is not valid within the FIPS provider), we *must* always have a `BN_CTX`.

mattcaswell commented:
I've rebased this to resolve a merge conflict with master. I've also fixed the indentation issues noted above, and added the `if (ctx == NULL` check also discussed above.

Please take another look.

mattcaswell commented:
Pushed. Thanks.
